webrtc/dtls/dtls_transport.nim

```diff
@@ -0,0 +1,179 @@
+#  * MIT license ([LICENSE-MIT](LICENSE-MIT))
+# at your option.
+# This file may not be copied, modified, or distributed except according to
+# those terms.
+
+import times, deques, tables, sequtils
+import chronos, chronicles
+import ./[dtls_utils, dtls_connection], ../errors,
+       ../stun/[stun_connection, stun_transport]
+
+import mbedtls/ssl
+import mbedtls/ssl_cookie
+import mbedtls/ssl_cache
+import mbedtls/pk
+import mbedtls/md
+import mbedtls/entropy
+import mbedtls/ctr_drbg
+import mbedtls/rsa
+import mbedtls/x509
+import mbedtls/x509_crt
+import mbedtls/bignum
+import mbedtls/error
+import mbedtls/net_sockets
+import mbedtls/timing
+
+logScope:
+  topics = "webrtc dtls"
+
+# Implementation of a DTLS client and a DTLS Server by using the Mbed-TLS library.
+# Multiple things here are unintuitive partly because of the callbacks
+# used by Mbed-TLS and that those callbacks cannot be async.
+
+# -- Dtls --
+
+type
+  Dtls* = ref object of RootObj
+    connections: Table[TransportAddress, DtlsConn]
+    transport: Stun
+    laddr*: TransportAddress
+    started: bool
+    ctr_drbg: mbedtls_ctr_drbg_context
+    entropy: mbedtls_entropy_context
+
+    serverPrivKey: mbedtls_pk_context
+    serverCert: mbedtls_x509_crt
+    localCert: seq[byte]
+
+proc new*(T: type Dtls, transport: Stun): T =
+  var self = T(
+    connections: initTable[TransportAddress, DtlsConn](),
+    transport: transport,
+    laddr: transport.laddr,
+    started: true
+  )
+
+  mb_ctr_drbg_init(self.ctr_drbg)
+  mb_entropy_init(self.entropy)
+  mb_ctr_drbg_seed(self.ctr_drbg, mbedtls_entropy_func, self.entropy, nil, 0)
+
+  self.serverPrivKey = self.ctr_drbg.generateKey()
+  self.serverCert = self.ctr_drbg.generateCertificate(self.serverPrivKey)
+  self.localCert = newSeq[byte](self.serverCert.raw.len)
+  copyMem(addr self.localCert[0], self.serverCert.raw.p, self.serverCert.raw.len)
+  return self
+
+proc stop*(self: Dtls) {.async.} =
+  if not self.started:
+    warn "Already stopped"
+    return
+
+  await allFutures(toSeq(self.connections.values()).mapIt(it.close()))
+  self.started = false
+
+# -- Remote / Local certificate getter --
+
+proc localCertificate*(self: Dtls): seq[byte] =
+  self.localCert
+
+# -- Dtls Accept / Connect procedures --
+
+proc cleanupDtlsConn(self: Dtls, conn: DtlsConn) {.async.} =
+  # Waiting for a connection to be closed to remove it from the table
+  await conn.join()
+  self.connections.del(conn.raddr)
+
+proc accept*(self: Dtls): Future[DtlsConn] {.async.} =
+  ## Accept a Dtls Connection
+  ##
+  var res = DtlsConn.new(await self.transport.accept(), self.laddr)
+
+  mb_ssl_init(res.ctx.ssl)
+  mb_ssl_config_init(res.ctx.config)
+  mb_ssl_cookie_init(res.ctx.cookie)
+  mb_ssl_cache_init(res.ctx.cache)
+
+  res.ctx.ctr_drbg = self.ctr_drbg
+  res.ctx.entropy = self.entropy
+
+  res.ctx.pkey = self.serverPrivKey
+  res.ctx.srvcert = self.serverCert
+  res.localCert = self.localCert
+
+  mb_ssl_config_defaults(
```

Comment:
There is some duplication between this code and the `connect` proc below. Maybe the common code could be moved to the `DtlsConn` constructor.

Reply:
Done